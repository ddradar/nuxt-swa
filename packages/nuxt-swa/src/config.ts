import type { NuxtConfig } from 'nuxt/config'

export const preConfiguredProviders = ['aad' as const, 'github' as const]
const knownIdentityProviders = [
  'apple',
  'facebook',
  'google',
  'twitter',
] as const

export interface ModuleOptions {
  /**
   * Authentication Provider list to login your app.
   * If your app does not use built-in Authentication/Authorization, please set this to `[]`.
   * @default ['aad', 'github']
   */
  authProviders?: (
    | (typeof knownIdentityProviders)[number]
    | Omit<string, (typeof knownIdentityProviders)[number]>
  )[]
  /**
   * Define custom roles to use in your app besides the default roles (`anonymous`, `authenticated`).
   * @default []
   */
  customRoles: Omit<string, 'anonymous' | 'authenticated'>[]
  /** Data API config */
  dataApi:
    | {
        /** REST endpoint path (same value as `runtime.rest.path` in `staticwebapp.database.config.json`) */ rest: string
        /** GraphQL endpoint path (same value as `runtime.graphql.path` in `staticwebapp.database.config.json`) */ graphql: string
      }
    | false
}
export const defaults: ModuleOptions = {
  customRoles: [],
  dataApi: { rest: '/rest', graphql: '/graphql' },
}

/** Resolve `authProviders` option from Nitro config. */
export function resolveAuthProviders(
  options: Required<Required<NuxtConfig>['nitro']>['azure']['config']
): Required<ModuleOptions>['authProviders'] {
  if (!options?.auth?.identityProviders) return preConfiguredProviders

  const identityProviders = options?.auth?.identityProviders

  const result: Required<ModuleOptions>['authProviders'] = []

  addPreConfiguredProviderIfEnabled('aad') // Microsoft Entra ID (formerly Azure Active Directory)
  addCustomProviderIfEnabled('apple') // Apple
  addCustomProviderIfEnabled('facebook') // Facebook
  addPreConfiguredProviderIfEnabled('github') // GitHub
  addCustomProviderIfEnabled('google') // Google
  addCustomProviderIfEnabled('twitter') // X (formerly Twitter)

  // OpenID Connect
  for (const key in identityProviders.customOpenIdConnectProviders) {
    if (identityProviders.customOpenIdConnectProviders[key]?.enabled !== false)
      result.push(key)
  }

  return result

  function addPreConfiguredProviderIfEnabled(
    provider: (typeof preConfiguredProviders)[number]
  ) {
    const key = provider === 'aad' ? 'azureActiveDirectory' : provider
    if (identityProviders[key]?.enabled !== false) result.push(provider)
  }

  function addCustomProviderIfEnabled(
    provider: (typeof knownIdentityProviders)[number]
  ) {
    if (
      identityProviders[provider] &&
      identityProviders[provider]?.enabled !== false
    )
      result.push(provider)
  }
}

export interface ModulePublicRuntimeConfig {
  /** Nuxt SWA config */
  swa: {
    /** REST API endpoint */ rest: string
    /** GraphQL endpoint */ graphql: string
  }
}

type SpecifiedRole = Exclude<UserRole, 'anonymous'>

declare module '#app' {
  interface PageMeta {
    /** List of roles required to view this page */
    allowedRoles?: SpecifiedRole | SpecifiedRole[]
  }
}

declare module 'vue-router' {
  interface RouteMeta {
    /** List of roles required to view this page */
    allowedRoles?: SpecifiedRole | SpecifiedRole[]
  }
}
