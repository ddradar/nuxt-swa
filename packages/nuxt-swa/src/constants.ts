// Azure Static Web Apps
export const _authBasePath = '/.auth'
export const _clientPrincipalHeader = 'x-ms-client-principal'

// My App
export const _authDataKey = 'swa-principal'
