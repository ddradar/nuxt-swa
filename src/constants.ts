export const _authBasePath = '/.auth'
export const _clientPrincipalHeader = 'x-ms-client-principal'
