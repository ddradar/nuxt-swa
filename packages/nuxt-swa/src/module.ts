import {
  defineNuxtModule,
  createResolver,
  addTypeTemplate,
  addImports,
} from '@nuxt/kit'

import { defaults, resolveAuthProviders } from './config'

export default defineNuxtModule({
  meta: {
    name: 'nuxt-swa',
    configKey: 'swa',
    compatibility: {
      nuxt: '^3.9.0',
    },
  },
  defaults,
  setup(options, nuxt) {
    const resolver = createResolver(import.meta.url)

    if (options.authProviders === undefined)
      options.authProviders = resolveAuthProviders(
        nuxt.options.nitro?.azure?.config
      )

    addTypeTemplate({
      filename: 'types/nuxt-swa.d.ts',
      src: resolver.resolve('runtime/types/nuxt-swa.d.ts.template'),
      options,
    })
    if (options.authProviders.length) {
      // Use auth feature
      addImports({
        name: 'useEasyAuth',
        from: resolver.resolve('runtime/composables/useEasyAuth'),
      })
    }
    // Add `nitro.azure.config` Typing
    addTypeTemplate({
      filename: 'types/AzureOptions.d.ts',
      src: resolver.resolve('runtime/types/AzureOptions.d.ts'),
    })
  },
})
