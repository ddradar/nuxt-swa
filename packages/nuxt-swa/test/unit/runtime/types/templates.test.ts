import { readFile } from 'node:fs/promises'
import { resolve } from 'node:path'

import { template } from 'lodash'
import { describe, expect, it } from 'vitest'

import { defaults, type SWAModuleOptions } from '~/src/module'

describe('runtime/types', () => {
  const minimumOptions: SWAModuleOptions = {
    authProviders: [],
    customRoles: [],
  }
  const fullOptions: SWAModuleOptions = {
    authProviders: ['github', 'twitter', 'line'],
    customRoles: ['administrator'],
  }
  describe.each(['UserRole.d.ts.template', 'IdentityProvider.d.ts.template'])(
    '%s',
    async file => {
      const pathes = [__dirname, '../../../../', 'src/runtime/types', file]
      const templateString = await readFile(resolve(...pathes), {
        encoding: 'utf-8',
      })

      it.each([defaults, minimumOptions, fullOptions])(
        '%o renders expected Type Definition',
        options =>
          expect(template(templateString)({ options })).toMatchSnapshot()
      )
    }
  )
})
