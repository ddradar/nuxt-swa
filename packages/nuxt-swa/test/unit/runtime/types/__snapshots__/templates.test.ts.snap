// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`runtime/types > IdentityProvider.d.ts.template > { authProviders: [ 'aad', 'github' ], customRoles: [] } renders expected Type Definition 1`] = `
"// Generated by nuxt-swa
type IdentityProvider =
  | 'aad'
  | 'github'
"
`;

exports[`runtime/types > IdentityProvider.d.ts.template > { authProviders: [ 'github', 'twitter', 'line' ], customRoles: [ 'administrator' ] } renders expected Type Definition 1`] = `
"// Generated by nuxt-swa
type IdentityProvider =
  | 'github'
  | 'twitter'
  | 'line'
"
`;

exports[`runtime/types > IdentityProvider.d.ts.template > { authProviders: [], customRoles: [] } renders expected Type Definition 1`] = `
"// Generated by nuxt-swa
type IdentityProvider =
"
`;

exports[`runtime/types > UserRole.d.ts.template > { authProviders: [ 'aad', 'github' ], customRoles: [] } renders expected Type Definition 1`] = `
"// Generated by nuxt-swa
type UserRole =
  | 'anonymous'
  | 'authenticated'

"
`;

exports[`runtime/types > UserRole.d.ts.template > { authProviders: [ 'github', 'twitter', 'line' ], customRoles: [ 'administrator' ] } renders expected Type Definition 1`] = `
"// Generated by nuxt-swa
type UserRole =
  | 'anonymous'
  | 'authenticated'
  | 'administrator'
"
`;

exports[`runtime/types > UserRole.d.ts.template > { authProviders: [], customRoles: [] } renders expected Type Definition 1`] = `
"// Generated by nuxt-swa
type UserRole =
  | 'anonymous'
  | 'authenticated'

"
`;
